### PR TITLE
Conditional string encoding for HTML shell

### LINKED ISSUE #5884: emcc.py throws if HTML shell contains Unicode character
For instance, try insering a `→` character into an HTML shell. On compilation, that will throw with something similar to:

```bash
Traceback (most recent call last):
  File "/srv/emsdk/emscripten/incoming/emcc.py", line 2725, in <module>
    run()
  File "/srv/emsdk/emscripten/incoming/emcc.py", line 1854, in run
    memfile, optimizer)
  File "/srv/emsdk/emscripten/incoming/emcc.py", line 2613, in generate_html
    html.write(html_contents.encode('utf-8'))
UnicodeDecodeError: 'ascii' codec can't decode byte 0xe2 in position 3119: ordinal not in range(128)
```

### PR DESCRIPTION
**Pull request by saschanaz**
Fixes #5884

### CODE REVIEW

File: tests/test_browser.py [hunk 1]
```diff
@@ -3718,3 +3720,14 @@ def test_access_file_after_heap_resize(self):
...
     self.run_browser('page.html', 'hello from file', '/report_result?15')
+
+  def test_unicode_html_shell(self):
+    open(self.in_dir('main.cpp'), 'w').write(self.with_report_result(r'''
+      int main() {
+        REPORT_RESULT(0);
+        return 0;
+      }
+    '''))
+    open(self.in_dir('shell.html'), 'w').write(open(path_from_root('src', 'shell.html')).read().replace('Emscripten-Generated Code', 'Emscripten-Generated Emoji 😅'))
```
Comment (kripken) [inline tests/test_browser.py]: lol

**Comment by saschanaz** [comment]
>as something we should use in all the places we currently have things like x.encode('utf-8')

This line is added by me for Py3 compat, which accidentally broke Py2 one. Other existing `.encode('utf-8')` should still be discussed in #5814 as they are only for Py2.

File: tests/test_browser.py (hunk 1 above)
Comment (saschanaz) [reply tests/test_browser.py]: 🤣

**Comment by kripken** [comment]
Hmm, I still don't see. Why is this line different in that way?

**Comment by saschanaz** [comment]
`html.write()` here always receives `bytes` as we opened `html` with `wb` mode, so I added `.encode('utf-8')` in a previous PR to convert Py3 Unicode string to `bytes`.

The simple `.encode('utf-8')` works perfectly in Py3, but in Py2 it is essentially equal to `.decode('ascii').encode('utf-8')` as the original string is already `bytes`, which causes ASCII range error like #5884 when the original includes non-ASCII bytes. Try this minimal repro:

```py
print('\xAC\x00'.encode('utf-8'))
'''
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
UnicodeDecodeError: 'ascii' codec can't decode byte 0xac in position 0: ordinal not in range(128)
'''
```

**Comment by kripken** [comment]
Great, thanks, now I see.

But wow, encoding and utf is tricky with these Python 2/3 differences...

**Comment by kripken** [comment]
lgtm, but there's a merge conflict here.

File: tools/shared.py [hunk 2]
```diff
@@ -2536,6 +2536,13 @@ def make_shared_library(js_file, wasm_file):
     f.close()
     return wso
 
+def asbytes(s):
+  if str is bytes:
```
Comment (asRIA) [inline tools/shared.py]: Shouldn't it be `if s is bytes` ?

Comment (saschanaz) [reply tools/shared.py]: `str is bytes` checks whether the current Python version uses byte string by default.

**Pull request merged**